test/backends.py

```diff
@@ -0,0 +1,217 @@
+
+                result = self.usr_np.empty_like(res_list, dtype=object,
+                                                shape=(len(res_list),))
+
+                # 'result[:] = res_list' may look tempting, however:
+                # https://github.com/numpy/numpy/issues/16564
+                for idx in range(len(res_list)):
+                    result[idx] = res_list[idx]
+
+                return result
+
+            def _pyro_norm(self, argument, normord):
+                """This works around numpy.linalg norm not working with scalars.
+
+                If the argument is a regular ole number, it uses :func:`numpy.abs`,
+                otherwise it uses ``usr_np.linalg.norm``.
+                """
+                # Wrap norm for scalars
+                from numbers import Number
+                if isinstance(argument, Number):
+                    return self.usr_np.abs(argument)
+                # Needed to play nicely with Jax, which frequently creates
+                # arrays of shape () when handed numbers
+                if (isinstance(argument, self.usr_np.ndarray)
+                        and argument.shape == ()):
+                    return self.usr_np.abs(argument)
+                return self.usr_np.linalg.norm(argument, normord)
+
+        return JaxPyroWrapper
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.python.gen_thermochem_code(sol=sol)
+
+    @staticmethod
+    def extract_interface(module, mechname: str, user_np):
+        base_cls = module.Thermochemistry
+
+        if user_np == jnp:
+            base_cls = PythonBackend._gen_jax_wrapper(base_cls)
+
+        return base_cls(user_np)
+
+
+class CppBackend(Backend):
+    def get_name():
+        return "cpp"
+
+    def fix_arg(self, x):
+        if isinstance(x, Iterable) and not isinstance(x, str):
+            return self.pyro.VectorDouble(x)
+        return super().fix_arg(x)
+
+    def fix_result(self, result):
+        if isinstance(result, self.pyro.VectorDouble):
+            return np.array(result)
+        return super().fix_result(result)
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.cpp.gen_thermochem_code(sol=sol, namespace=name)
+
+
+class FortranBackend(Backend):
+    def get_name():
+        return "fortran90"
+
+    @staticmethod
+    def extract_interface(module, mechname: str, user_np):
+        return getattr(module, f"libpyro_fortran90_{mechname}")
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.fortran90.gen_thermochem_code(
+            sol=sol, module_name=name
+        )
+
+
+BACKENDS: typing.Dict[str, Backend] = {
+    CppBackend.get_name(): CppBackend,
+    FortranBackend.get_name(): FortranBackend,
+    PythonBackend.get_name(): PythonBackend
+}
+
+
+class Pyrometheus:
```

Comment:
Please add a docstring for this class. I am still struggling to understand what role it plays. It seems to be a wrapper for _all_ backends?

Reply:
Added the docstrings.